### DISCUSSION TITLE
Which files need to be persisted to keep WebUI configuration changes through container restarts?

### DISCUSSION THREAD

**Discussion by aq-daniel** [RaspAP/raspap-docker#43] [Q&A]
Thank you for your work and providing a containerized version of RaspAP. 👍 It's a good solution to use containerized RaspAP, if you have other services running beside it on your host device.

Now to my current problem:
I can successfully set some settings through the environment variables in the docker-compose.yml like RASPAP_SSID, WEBGUI_PORT, etc.
That's good and is working, but if I change any settings (e.g. Wireless Mode) or add for example an opnvpn config-file through the webui, it doesn't persist container restarts.

The correct solution for this problem seems to be a docker volume as mentioned in another discussion [here](https://github.com/RaspAP/raspap-docker/discussions/37#discussioncomment-9017184).
So my questions are:
1. Which files need to be mounted to a docker volume to persist setting changes through the WebUI over container restarts? Is there a list of all config files somewhere?
2. Does a config folder exist in the container, which could be bound to a volume? [In the docs](https://docs.raspap.com/defaults/) is a config directory mentioned, but I couldn't find it inside the docker container. 
3. I tried binding /var/www/html/includes to a docker volume as this directory contains the config.php, but it doesn't seem to work for persisting the wirelss mode setting over restarts. Which file contains the setting for the wireless mode?

**Comment by billz** [comment]
Hi, great questions. All of the file paths used by the application are defined in [config.php](https://github.com/RaspAP/raspap-webgui/blob/master/config/config.php). Persisting these to a docker volume will get you most of the way there. Note that the individual files defined here, not just `config.php` itself, must be saved to a storage volume if your goal is to persist all application settings through container restarts. RaspAP stores user preferences in `/etc/raspap`, so you could persist the contents of this directory as well.

Hope this gets you started in the right directions, and please do share your results! We can look at adding steps in the official docs to assist other users.

**Reply by fheinz** [reply]
I'm pursuing this avenue (and will contribiute my results). Is there a reason why not to mount the whole of `/etc` from a directory in the host?

**Reply by octavioj** [reply]
I am looking to do this as well because when I tried to switch the container to Bridged AP mode and tried to save it, I got "Unable to save WIFI Hotspot settings" so I wanted to have this saved in a volume to make sure it could be saved. If I read this thread correctly, for each instance of the container there needs be two volumes, one to bind to /etc/raspap and another to bind to /var/www/html/includes . Is that the correct understanding?

**Reply by billz** [reply]
One volume for `/etc/raspap` (this is where user preferences are stored), plus one mount per underlying service config file that `config.php` points to. So, append the following to the end of `docker-compose.yaml`:

```
   volumes:
      - /sys/fs/cgroup:/sys/fs/cgroup:rw
      - raspap-config:/etc/raspap
      - ./hostapd.conf:/etc/hostapd/hostapd.conf
      - ./dhcpcd.conf:/etc/dhcpcd.conf
      - ./wpa_supplicant.conf:/etc/wpa_supplicant/wpa_supplicant.conf
      - ./dnsmasq.d:/etc/dnsmasq.d
    restart: unless-stopped

volumes:
  raspap-config:
```
Hope this helps!

**Reply by octavioj** [reply]
Perfect, thanks! Will try now.
